File: solutions/container_with_most_water/container_with_most_water.py
def maxArea(height: List[int]) -> int:
    left, right = 0, len(height) - 1
    max_area = 0

    while left < right:
        # width between lines
        width = right - left
        # height is the smaller line
        current_height = min(height[left], height[right])
        # area = width * height
        max_area = max(max_area, width * current_height)

        # move the pointer of the smaller line
        if height[left] < height[right]:
            left += 1
        else:
            right -= 1

    return max_area

# Test
print(maxArea([1,8,6,2,5,4,8,3,7]))  # Expected output: 49
print(maxArea([1,1]))                # Expected output: 1

File: solutions/container_with_most_water/container_with_most_water.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int maxArea(vector<int>& height) {
    int left = 0, right = height.size() - 1;
    int max_area = 0;

    while (left < right) {
        int width = right - left;
        int current_height = min(height[left], height[right]);
        max_area = max(max_area, width * current_height);

        if (height[left] < height[right])
            left++;
        else
            right--;
    }
    return max_area;
}

int main() {
    vector<int> heights1 = {1,8,6,2,5,4,8,3,7};
    cout << "Max Area: " << maxArea(heights1) << endl;  // Expected: 49

    vector<int> heights2 = {1,1};
    cout << "Max Area: " << maxArea(heights2) << endl;  // Expected: 1

    return 0;
}
